### PR TITLE
Chord-Camber Distance: signed distance from chord line per surface node

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model currently has no explicit encoding of **upper vs lower surface** or **camber** for each surface node. For NACA0012 (symmetric, in-distribution), upper and lower surface are mirror images. For NACA6416 (cambered, OOD in p_tan), the pressure distribution is fundamentally asymmetric due to camber — the suction peak is stronger and shifted forward on the upper surface.

By adding a single input channel encoding each surface node's **signed distance from the chord line** (LE-to-TE straight line), the model receives explicit upper/lower surface discrimination that is geometry-invariant:
- NACA0012: values are symmetric around zero (thin, no camber)
- NACA6416: values are asymmetric — upper surface nodes have larger positive distances due to 6% camber

This is NOT derivable from existing features. The DSDF features encode distance TO the surface from volume nodes, not ALONG the surface geometry. The TE/LE coordinate frames encode distance FROM reference points, not the surface's relationship to its own chord line.

**Physical motivation:** Aerodynamicists parameterize airfoil surfaces as `y_upper(x/c)` and `y_lower(x/c)` relative to the chord line. The signed chord distance IS the thickness distribution that determines the pressure distribution. For cambered foils, the asymmetry of this distance is what creates the camber-induced lift.

**Expected improvement:** -1 to -3% p_tan (OOD camber discrimination), possible p_oodc improvement (better geometry understanding).

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flag

```python
chord_camber_dist: bool = False   # signed distance from chord line per surface node (+1 channel)
```

### Step 2: Implement `compute_chord_camber_distance`

Add after the existing feature computation functions:

```python
def compute_chord_camber_distance(raw_xy, is_surface, saf_norm):
    """Signed distance from chord line (LE-to-TE) for each surface node.
    
    For each foil: find LE (min-x surface node) and TE (max-x surface node),
    compute the chord line direction, then project each node's offset from LE
    onto the normal of the chord line. Normalize by chord length.
    
    Args:
        raw_xy:     [B, N, 2] raw x, y coordinates
        is_surface: [B, N] bool
        saf_norm:   [B, N] saf channel norm (<=0.005 = foil-1, >0.005 = foil-2)
    
    Returns: [B, N, 1] signed chord-normal distance, normalized by chord length.
             Positive = above chord line, negative = below. Zero for volume nodes.
    """
    B, N = raw_xy.shape[:2]
    x_coords = raw_xy[:, :, 0]
    y_coords = raw_xy[:, :, 1]
    INF = 1e6
    result = torch.zeros(B, N, 1, device=raw_xy.device, dtype=raw_xy.dtype)
    
    for foil_idx, threshold_fn in enumerate([
        lambda sn: sn <= 0.005,   # fore-foil
        lambda sn: sn > 0.005,    # aft-foil
    ]):
        mask = is_surface & threshold_fn(saf_norm)
        if foil_idx == 1:
            # Skip aft-foil for single-foil samples
            has_aft = mask.any(dim=1)
            if not has_aft.any():
                continue
        
        # Find LE (min-x) and TE (max-x) among this foil's surface nodes
        x_masked_min = x_coords.clone()
        x_masked_min[~mask] = INF
        le_idx = x_masked_min.argmin(dim=1)  # [B]
        
        x_masked_max = x_coords.clone()
        x_masked_max[~mask] = -INF
        te_idx = x_masked_max.argmax(dim=1)  # [B]
        
        le_x = x_coords.gather(1, le_idx.unsqueeze(1)).squeeze(1)  # [B]
        le_y = y_coords.gather(1, le_idx.unsqueeze(1)).squeeze(1)
        te_x = x_coords.gather(1, te_idx.unsqueeze(1)).squeeze(1)
        te_y = y_coords.gather(1, te_idx.unsqueeze(1)).squeeze(1)
        
        # Chord direction and normal
        chord_dx = te_x - le_x  # [B]
        chord_dy = te_y - le_y
        chord_len = (chord_dx**2 + chord_dy**2).sqrt().clamp(min=1e-6)  # [B]
        
        # Normal direction (perpendicular to chord, pointing "up")
        # n = (-chord_dy, chord_dx) / chord_len
        nx = -chord_dy / chord_len  # [B]
        ny = chord_dx / chord_len
        
        # Signed distance from chord line for each node
        dx_from_le = x_coords - le_x[:, None]  # [B, N]
        dy_from_le = y_coords - le_y[:, None]
        signed_dist = dx_from_le * nx[:, None] + dy_from_le * ny[:, None]  # [B, N]
        signed_dist = signed_dist / chord_len[:, None]  # normalize by chord
        
        # Zero out non-surface and non-this-foil nodes
        signed_dist = signed_dist * mask.float()
        result[:, :, 0] = result[:, :, 0] + signed_dist
    
    return result
```

### Step 3: Update input dimension

Where `fun_dim` or `n_x` is computed, add:
```python
+ (1 if cfg.chord_camber_dist else 0)
```

### Step 4: Compute and append in training/eval loops

Before normalization, save raw_xy. Then after the existing feature concatenation:
```python
if cfg.chord_camber_dist:
    camber_feat = compute_chord_camber_distance(_raw_xy, is_surface, _raw_saf_norm)
    x = torch.cat([x, camber_feat], dim=-1)
```

Apply in **all 4 loops** (train, val, vis, re-verification).

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/chord-camber-s42" \
  --wandb_group "round17/chord-camber-distance" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --chord_camber_dist

# Seed 73 — identical but --seed 73 --wandb_name "frieren/chord-camber-s73"
```

### Step 6: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, comparison to baseline, W&B run IDs.

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: hgml7i2r (seed 42), qic03vrg (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (mae_surf_p)

| Metric | Baseline (PR #2213) | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|---------------------|---------|---------|------------|-------|
| **p_in** | **11.979** | 12.828 | 12.443 | 12.636 | **+5.5%** |
| **p_tan** | **28.341** | 28.914 | 28.304 | 28.609 | **+0.9%** |
| **p_oodc** | **7.643** | 7.848 | 7.891 | 7.870 | **+3.0%** |
| **p_re** | **6.300** | 6.556 | 6.741 | 6.649 | **+5.5%** |

### W&B Run IDs
- Seed 42: `zs57oi9m` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zs57oi9m)
- Seed 73: `xfhmmdcg` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xfhmmdcg)

### Command used
```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/chord-camber-s42" \
  --wandb_group "round17/chord-camber-distance" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --chord_camber_dist
```
(Seed 73 identical with `--seed 73 --wandb_name "frieren/chord-camber-s73"`)

Peak memory: ~60GB (seed 42), ~76GB (seed 73)

### What happened

**Negative result.** The chord-camber distance feature hurt performance across all metrics, with the largest regressions in p_in (+5.5%) and p_re (+5.5%).

**Why it likely failed:**

1. **Redundancy with existing features.** The signed chord distance for each surface node is closely related to information already encoded in the DSDF channels. The DSDF gradients at surface nodes implicitly encode surface normal direction, which correlates with the chord-normal distance. The model can already distinguish upper/lower surface from the DSDF gradient orientation relative to the flow.

2. **Noise on single-foil (in-distribution) samples.** For symmetric NACA0012 at varying AoA, the chord line rotates relative to the flow direction but the feature is computed in the geometry frame. The signed distance may confuse the model by providing upper/lower labels that flip meaning as AoA changes, since what matters physically is suction/pressure side (flow-relative), not geometric upper/lower.

3. **One extra input channel.** Even though it's a single channel, adding a feature the model can't usefully correlate with pressure differences means the extra dimension adds noise that regularization must overcome. The +5.5% in-distribution regression suggests the feature actively interferes with learning.

4. **LE/TE detection noise.** For meshes with blunt trailing edges, the max-x heuristic for TE detection may pick slightly different nodes across samples, adding micro-noise to the chord line definition.

### Suggested follow-ups

- **Flow-relative surface parameterization** instead of geometry-relative: encode suction/pressure side based on the stagnation point location (where surface velocity ≈ 0), which adapts to AoA. This requires knowing the flow solution though, so it's only useful as a self-supervised auxiliary target.
- **Normalized arc-length along surface** (0 at LE, 1 at TE on both upper and lower surfaces) might be more useful — it provides a canonical parameterization without the upper/lower ambiguity issue.
- **Drop this idea** — the DSDF features already encode sufficient geometric information about surface shape. The model's p_tan bottleneck is more likely about wake coupling and inter-foil interaction (where wake deficit feature helped) than about surface geometry encoding.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2227 because ADVISOR: Closing as dead end. All metrics regressed: p_in +5.5%, p_re +5.5%, p_oodc +3.0%, p_tan +0.9%.

Your analysis is excellent — the core failure is that geometry-frame upper/lower does NOT equal flow-frame suction/pressure. At nonzero AoA, the signed chord distance provides misleading labels that flip meaning as AoA changes. Additionally, DSDF gradient orientation already partially encodes surface normal direction, making this feature partially redundant.

Key takeaway: for surface features to help, they need to be either (a) in flow-relative coordinates (AoA-aware), or (b) encoding genuinely new shape information not derivable from DSDF. Moving you to a new experiment..
